src/NuGet.Core/NuGet.Protocol/Resources/PackageMetadataResourceV3.cs

```diff
@@ -38,37 +55,240 @@ public class PackageMetadataResourceV3 : PackageMetadataResource
+            ILogger log,
+            CancellationToken token)
+        {
+            var metadataCache = new MetadataReferenceCache();
+            var registrationUri = _regResource.GetUri(packageId);
+
+            var registrationIndexResult = await LoadRangesAsync(
+                _client,
+                registrationUri,
+                packageId,
+                sourceCacheContext,
+                async httpSourceResult => await ProcessRegistrationIndexAsync<RegistrationIndex>(httpSourceResult, token),
+                log,
+                token);
+
+            var registrationIndex = registrationIndexResult.Index;
+            var httpSourceCacheContext = registrationIndexResult.CacheContext;
+
+            if (registrationIndex == null)
+            {
+                // The server returned a 404, the package does not exist
+                return Enumerable.Empty<PackageSearchMetadataRegistration>();
+            }
+
+            var results = new List<PackageSearchMetadataRegistration>();
+
+            foreach (var registrationPage in registrationIndex.Items)
+            {
+                if(registrationPage==null)
+                {
+                    throw new InvalidDataException(registrationUri.AbsoluteUri);
+                }
+
+                var lower = NuGetVersion.Parse(registrationPage.Lower);
+                var upper = NuGetVersion.Parse(registrationPage.Upper);
+
+                if (range.IsItemRangeRequired(lower, upper))
+                {
+                    if (registrationPage.Items == null)
+                    {
+                        var rangeUri = registrationPage.Url;
+                        var leafRegistrationPage = await GetRegistratioIndexPageAsync(_client, rangeUri, packageId, lower, upper, httpSourceCacheContext, log, token);
+
+                        if(registrationPage==null)
+                        {
+                            throw new InvalidDataException(registrationUri.AbsoluteUri);
+                        }
+
+                        ProcessRegistrationPage(leafRegistrationPage, results, range, includePrerelease,includeUnlisted, metadataCache);
+                    }
+                    else
+                    {
+                        ProcessRegistrationPage(registrationPage, results, range, includePrerelease, includeUnlisted, metadataCache);
+                    }
+                }
+            }
+
+            return results;
+        }
+
+        /// <summary>
+        /// Process RegistrationIndex and return list of RegistrationPages.
+        /// </summary>
+        /// <typeparam name="T">Generic type</typeparam>
+        /// <param name="httpSourceResult">HttpSourceResult type object which comes from HttpSource, used for get stream.</param>
+        /// <param name="token">Cancellation token.</param>
+        /// <returns></returns>
+        internal async Task<T> ProcessRegistrationIndexAsync<T>(HttpSourceResult httpSourceResult, CancellationToken token)
```

Comment:
```suggestion
        private Task<T> Deserialize<T>(HttpSourceResult httpSourceResult, CancellationToken token)
```

"Process" is a very generic word, but makes me think of doing some kind of computation or aggregation. Plus this method is used not just for the index, but pages as well.

Also looks like it's never used outside this class.

Also isn't async. Uses `Task.FromResult`, while still using the `async` keyword on the method, thereby losing any benefit of `Task.FromResult`

Reply:
I renamed to DeserializeStreamDataAsync, because in the future we might move out to common usage outside this class to deserialize `Stream` type. 

But want to keep it async, because it passed down to below method need `Task<T>` in `HttpSource.cs`. 
One solution is I copy this method and create another without `Task` but it'll result another 140 lines of code inside `HttpSource.cs`.

![image](https://user-images.githubusercontent.com/8766776/85185855-b9a2ce00-b24a-11ea-83e3-40c4ae456538.png)